crates/runtime-rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,400 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{sync::Arc, time::Duration};
+
+use governor::{
+    Jitter, Quota, RateLimiter as Governor,
+    clock::DefaultClock,
+    middleware::NoOpMiddleware,
+    state::{InMemoryState, NotKeyed},
+};
+use rand::prelude::*;
+use snafu::ResultExt;
+use snafu::prelude::*;
+use tokio::sync::{Semaphore, SemaphorePermit};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Failed to acquire semaphore permit.\n{source}"))]
```

Comment:
```suggestion
    #[snafu(display("Failed to acquire semaphore permit. {source}"))]
```

The error guidelines haven't been updated yet, but let's start to remove newlines from error messages going forward.